### PR TITLE
release-21.2: randgen: do not create computed columns with bad volatilities

### PR DESCRIPTION
**Pull request by mgartner**
Some cast to STRING types has been given an incorrect volatility. For
example, REGCLASS->STRING casts are immutable when they should be stable
(see #74286 and #74553 for more details).

Creating computed column expressions with such a cast can cause logical
correctness bugs and internal errors. The volatilities cannot be fixed
without causing backward incompatibility. This commit prevents `randgen`
from creating computed columns with these casts so that sqlsmith and TLP
do not repetitively find these known volatility bugs.

Informs #74727

Release note: None

Release justification: This is a test-only change.

### CODE REVIEW

**Review by rytaft — APPROVED** [review]
:lgtm: Thanks!

> Reviewed 1 of 1 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/75151#-:-MtnCx6zA_8IdGn7e7IG:b-g2a5fb)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @mgartner)

**Review by rafiss — CHANGES_REQUESTED** [review]
ah nice, then we can probably clean this part of the code up: https://github.com/cockroachdb/cockroach/blob/efb424d2adaf443fc816ea607a135c8c1ac41339/pkg/sql/randgen/mutator.go#L842

File: pkg/sql/randgen/expr.go [hunk 1]
```diff
@@ -243,3 +244,24 @@ func randExpr(
...
+// typeToStringCastHasIncorrectVolatility returns true for a given type if the
+// cast from it to STRING types has been given an incorrect volatility. For
+// example, REGCLASS->STRING casts are immutable when they should be stable (see
+// #74286 and #74553 for more details).
+//
+// Creating computed column expressions with such a cast can cause logical
+// correctness bugs and internal errors. The volatilities cannot be fixed
+// without causing backward incompatibility, so this function is used to prevent
+// sqlsmith and TLP from repetitively finding these known volatility bugs.
+func typeToStringCastHasIncorrectVolatility(t *types.T) bool {
```
Comment (rafiss) [inline pkg/sql/randgen/expr.go:257]: should we also throw in TimestampFamily and DateFamily? (maybe IntervalFamily too, not sure)

Comment (mgartner) [reply pkg/sql/randgen/expr.go:257]: Good point. In that case we should merge this to master first then backport to the other releases. I'll close this and open a new PR.

Comment (mgartner) [reply pkg/sql/randgen/expr.go:257]: INTERVAL::STRING is immutable in Postgres, so I don't think we need that:

```
marcus=# CREATE TABLE t (i INTERVAL, t TEXT GENERATED ALWAYS AS (i::TEXT) STORED);
CREATE TABLE
```

Though I'm confused why this cast depends on `IntervalStyle`. It shouldn't if it's immutable.

https://github.com/cockroachdb/cockroach/blob/979d3ff35cfd6b163cb26b4c911c9bae78597d41/pkg/sql/sem/tree/cast.go#L573-L574

Comment (otan) [reply pkg/sql/randgen/expr.go:257]: > INTERVAL::STRING is immutable in Postgres

:blob_eyes_grin:

**Pull request closed without merging**